pyOCD/pyDAPAccess/dap_access_cmsis_dap.py

```diff
@@ -165,46 +164,77 @@ def _dap_transfer_block_free_bytes(self, size):
         recv = size - 4 - 4 * self._read_count
         return send, recv
 
-    def get_read_space(self):
+    def _get_free_read_transfer_words(self):
         """
         Return the number of words free in the response packet
         """
-        # Size left if useing DAP_Transfer
         send, recv = self._dap_transfer_free_bytes(self._size)
-        transfer_size = min(send, recv // 4)
-
-        # Size left if using DAP_TransferBlock
-        _, recv = self._dap_transfer_block_free_bytes(self._size)
-        transfer_block_size = recv // 4
-
-        # If data does not allow block transfer invalidate it
-        if not self._block_allowed:
-            transfer_block_size = -1
-
-        size = max(transfer_size, transfer_block_size)
-        assert size >= 0
-        return size
+        return min(send, recv // 4)
 
-    def get_write_space(self):
+    def _get_free_write_transfer_words(self):
         """
         Return the number of words free in the transmit packet
         """
-        # Size left if useing DAP_Transfer
         send, _ = self._dap_transfer_free_bytes(self._size)
-        transfer_size = send // 5
+        return send // 5
 
-        # Size left if using DAP_TransferBlock
+    def _get_free_read_transfer_block_words(self):
+        """
+        Return the number of words free in the response packet
+        """
+        _, recv = self._dap_transfer_block_free_bytes(self._size)
+        return recv // 4
+
+    def _get_free_write_transfer_block_words(self):
+        """
+        Return the number of words free in the transmit packet
+        """
         send, _ = self._dap_transfer_block_free_bytes(self._size)
-        transfer_block_size = (self._size - 5) // 4 - self._write_count
+        return send // 4
 
-        # If data does not allow block transfer invalidate it
-        if not self._block_allowed:
-            transfer_block_size = -1
+    def get_request_space(self, count, request, dap_index):
```

Comment:
You mean does `get_request_space()` need to return the size? Yes, it does. The returned size varies depending on the existing packet contents and new request. This allows for using packets most efficiently.

Below is a section of logging showing the request/add sequences for the beginning of the speed test where it writes to RAM.

Btw, should I leave in the debug log statements like this but with a flag to control enabling? (Like the logging flags at the top of gdbserver.py.)

```
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(1, 05:w)[wc=0, rc=0, ba=1->1] -> (sz=1, free=254)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(1, 05:w) -> [wc=1, rc=0, ba=1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(1024, 0d:w)[wc=1, rc=0, ba=1->0] -> (sz=203, free=203, delta=-51)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(203, 0d:w) -> [wc=204, rc=0, ba=0]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:_write: send packet [1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:New _Command
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(821, 0d:w)[wc=0, rc=0, ba=1->1] -> (sz=254, free=254)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(254, 0d:w) -> [wc=254, rc=0, ba=1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:_write: send packet [1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:New _Command
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(567, 0d:w)[wc=0, rc=0, ba=1->1] -> (sz=254, free=254)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(254, 0d:w) -> [wc=254, rc=0, ba=1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:_write: send packet [1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:New _Command
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(313, 0d:w)[wc=0, rc=0, ba=1->1] -> (sz=254, free=254)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(254, 0d:w) -> [wc=254, rc=0, ba=1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:_write: send packet [1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:New _Command
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(59, 0d:w)[wc=0, rc=0, ba=1->1] -> (sz=59, free=254)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(59, 0d:w) -> [wc=59, rc=0, ba=1]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(1, 05:w)[wc=59, rc=0, ba=1->0] -> (sz=1, free=145, delta=-195)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(1, 05:w) -> [wc=60, rc=0, ba=0]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:get_request_space(1024, 0d:w)[wc=60, rc=0, ba=0->0] -> (sz=144, free=144, delta=-51)
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:add(144, 0d:w) -> [wc=204, rc=0, ba=0]
DEBUG:pyOCD.pyDAPAccess.dap_access_cmsis_dap:_write: send packet [1]
```